mlir/lib/Dialect/SCF/Utils/Utils.cpp

```diff
@@ -1070,104 +1071,164 @@ TileLoops mlir::extractFixedOuterLoops(scf::ForOp rootForOp,
   return tileLoops;
 }
 
-scf::ForallOp mlir::fuseIndependentSiblingForallLoops(scf::ForallOp target,
-                                                      scf::ForallOp source,
-                                                      RewriterBase &rewriter) {
-  unsigned numTargetOuts = target.getNumResults();
-  unsigned numSourceOuts = source.getNumResults();
+//===----------------------------------------------------------------------===//
+// Fusion related helpers
+//===----------------------------------------------------------------------===//
 
-  // Create fused shared_outs.
-  SmallVector<Value> fusedOuts;
-  llvm::append_range(fusedOuts, target.getOutputs());
-  llvm::append_range(fusedOuts, source.getOutputs());
+bool mlir::checkFusionStructuralLegality(LoopLikeOpInterface &target,
+                                         LoopLikeOpInterface &source) {
+  auto iterSpaceEq =
+      target.getLoopLowerBounds() == source.getLoopLowerBounds() &&
+      target.getLoopUpperBounds() == source.getLoopUpperBounds() &&
+      target.getLoopSteps() == source.getLoopSteps();
+  auto forAllTarget = dyn_cast<scf::ForallOp>(*target);
+  auto forAllSource = dyn_cast<scf::ForallOp>(*source);
+  if (forAllTarget && forAllSource)
+    return iterSpaceEq &&
+           forAllTarget.getMapping() == forAllSource.getMapping();
```

Comment:
it is very undesirable to check specialized properties for concrete loops here, as this is supposed to work generically on `LoopLIkeOpInterface`.  Does it make sense to add a `getMapping` interface method to LoopLikeInterface.  Alternatively, this could be intended as a generic check, and other specialized checks would be done externally. Any thoughts?
@matthias-springer @ftynse @Hardcode84

Reply:
We can live with this, especially given that this is located under SCF. More interfacey solution could be to have a `bool isFusableWithSibling(LoopLikeOpInterface)` method on the interface that would have this logic, and potentially enable sibling fusion of different kinds of loops.

Reply:
ok. i think i will put a TODO though. i dont want to set a precedent that goes against one of the main goals of this PR (to unify common loop like things). i do believe there is a good abstraction to clean up this and the `fuseIndependentSiblingParallelLoops` function, it's just still a work in progress. I don't want to give up on that nor give the impression that it can't be done just because it wasn't finished in this PR